src/irc/Ident.ts

```diff
@@ -0,0 +1,160 @@
+    }
+
+    public configure(opts: IdentConfig) {
+        log.info("Configuring ident server => %s", JSON.stringify(opts));
+        this.config = opts;
+    }
+
+    public setMapping(username: string, port: number) {
+        if (port > 0) {
+            this.portMappings[port] = username;
+            log.debug("Set user %s on port %s", username, port);
+        }
+        else if (port === 0) {
+            Object.keys(this.portMappings)
+            .filter((portNum: string) => this.portMappings[portNum] === username)
+            .forEach((portNum) => {
+                if (this.portMappings[portNum] === username) {
+                    delete this.portMappings[portNum];
+                    log.debug("Remove user %s from port %s", username, portNum);
+                }
+            });
+        }
+    }
+
+    private onConnection(sock: net.Socket) {
+        log.debug("CONNECT %s %s", sock.remoteAddress, sock.remotePort);
+        sock.on("data", (data) => {
+            log.debug("DATA " + data);
+            const ports = data.toString().split(",");
+            const remoteConnectPort = Number(ports[1]);
+            const localOutgoingPort = Number(ports[0]);
+            if (!remoteConnectPort || !localOutgoingPort) {
+                log.debug("BAD DATA");
+                sock.end();
+                return;
+            }
+            this.tryRespond(sock,
+                String(localOutgoingPort),
+                String(remoteConnectPort)).catch((ex) => {
+                    // Just close the connection
+                    sock.end();
+                });
+        });
+        sock.on("close", () => {
+            log.debug("CLOSE");
+        });
+        sock.on("error", (err) => {
+            log.error("connection error: " + err);
+            if (err && err.stack) {
+                log.error(err.stack);
+            }
+        });
+    }
+
+    public clientBegin(): () => void { 
+        log.debug("IRC client started connection");
+        let res!: () => void;
+        const p: Promise<void> = new Promise((resolve) => {
+            res = resolve;
+            setTimeout(resolve, CLIENT_CONNECTION_TIMEOUT_MS);
+        });
+        this.pendingConnections.add(p);
+        p.then(() => {
+            log.debug("IRC client connected");
+            this.pendingConnections.delete(p);
+        })
+        return res;
+    }
+
+    private async tryRespond(sock: net.Socket, localPort: string, remotePort: string) {
+        const username = this.portMappings[localPort];
+        if (username) {
+            log.debug("Port %s is %s", localPort, username);
+            this.respond(sock, localPort, remotePort, username);
+            return;
+        }
+        // Wait for pending connections to finish first.
+        await Promise.all([...this.pendingConnections]);
+        log.debug("No user on port %s", localPort);
+        this.respond(sock, localPort, remotePort);
```

Comment:
You need to recheck the `portMappings` after waiting on the pending connections.